odl/trafos/convolution.py

```diff
@@ -0,0 +1,713 @@
+                new_dom_csides.append(ds)
+            else:
+                # Pick the coarser one if 'down' or the finer one if 'up'
+                if ((ks < ds and resamp == 'up') or
+                        (ks > ds and resamp == 'down')):
+                    new_ker_csides.append(ks)
+                    new_dom_csides.append(ks)
+                else:
+                    new_ker_csides.append(ds)
+                    new_dom_csides.append(ds)
+
+        return new_ker_csides, new_dom_csides
+
+    @property
+    def impl(self):
+        """Implementation of this operator."""
+        return self._impl
+
+    @property
+    def resample(self):
+        """Resampling used during evaluation."""
+        return self._resample
+
+    def kernel(self):
+        """Return the real-space kernel of this convolution operator.
+
+        Note that internally, the scaled kernel is stored if
+        ``scale=True`` (default) was chosen during initialization, i.e.
+        calculating the original kernel requires to reverse the scaling.
+        Otherwise, the original unscaled kernel is returned.
+        """
+        if self._scale:
+            return self._kernel / self._kernel_scaling
+        else:
+            return self._kernel
+
+    def _call(self, x):
+        """Implement ``self(x)``."""
+        if np.allclose(self._kernel.space.cell_sides, self.domain.cell_sides):
+            # No resampling needed
+            return signal.convolve(x, self._kernel, mode='same')
+        else:
+            if self.domain != self._domain_resampling_op.range:
+                # Only resample if necessary
+                x = self._domain_resampling_op(x)
+
+            if self._kernel.space != self._kernel_resampling_op.range:
+                kernel = self._kernel_resampling_op(self._kernel)
+            else:
+                kernel = self._kernel
+
+            conv = signal.convolve(x, kernel, mode='same')
+            return self._domain_resampling_op.inverse(conv)
+
+    @property
+    def adjoint(self):
+        """Adjoint operator defined by the reversed kernel."""
+        # Make space for the adjoint kernel, -S if S is the kernel space
+        adj_min_corner = -self._kernel.space.max_corner
+        adj_max_corner = -self._kernel.space.min_corner
+        adj_ker_space = uniform_discr_fromdiscr(
+            self._kernel.space,
+            min_corner=adj_min_corner, max_corner=adj_max_corner)
+
+        # Reverse the kernel
+        reverse_slice = (slice(None, None, -1),) * self.domain.ndim
+        adj_kernel = adj_ker_space.element(
+            self._kernel.asarray()[reverse_slice])
+
+        adj_conv = RealSpaceConvolution(self.range, adj_kernel, impl=self.impl,
+                                        scale=False, resample=self.resample)
+        # Kernel is already scaled, but we need to make sure that the
+        # kernel() method returns the rescaled kernel.
+        adj_conv._scale = True
```

Comment:
This is a slight hack, but I don't want to scale back and forth in a property, could be expensive.